AssertionError indexing snapshot not registered in DUOS

Indexing error on anvildev due to regression from the merge of PR #6872

**CloudWatch Logs Insights**    
region: us-east-1    
log-group-names: /aws/lambda/azul-indexer-anvildev-contribute_retry    
start-time: 2025-02-05T17:14:14.699Z    
end-time: 2025-02-05T17:23:11.088Z    
query-string:
  ```
  fields @timestamp, @message
| filter @requestId = 'bc916a15-8e8c-500e-899b-0a08111b5048'
| sort @timestamp asc
| limit 200
  ```
---
| @timestamp | @message |
| --- | --- |
| 2025-02-05 17:18:28.113 | START RequestId: bc916a15-8e8c-500e-899b-0a08111b5048 Version: $LATEST |
| 2025-02-05 17:18:28.113 | [INFO] 2025-02-05T17:18:28.113Z bc916a15-8e8c-500e-899b-0a08111b5048 azul.indexer.index_controller Worker handling message {'action': 'add', 'notification': {'transaction_id': '4b3447c7-0826-4fb3-8d8c-445b9c297da7', 'bundle_fqid': {'uuid': '677dd55c-3fa3-ab07-8c98-985d94d7577e', 'version': '2022-06-01T00:00:00.000000Z', 'source': {'id': '790795c4-49b1-4ac8-a060-207b92ea08c5', 'spec': 'tdr:bigquery:gcp:datarepo-dev-e53e74aa:ANVIL_1000G_2019_Dev_20230609_ANV5_202306121732:/0'}, 'table_name': 'anvil_dataset', 'batch_prefix': None}}, 'catalog': 'anvil'}, attempt #'1' (approx). |
| 2025-02-05 17:18:28.114 | [INFO] 2025-02-05T17:18:28.114Z bc916a15-8e8c-500e-899b-0a08111b5048 azul.plugins.repository.tdr_anvil Bundle '677dd55c-3fa3-ab07-8c98-985d94d7577e' is a DUOS bundle |
| 2025-02-05 17:18:28.128 | [INFO] 2025-02-05T17:18:28.128Z bc916a15-8e8c-500e-899b-0a08111b5048 azul.deployment Allocated new Boto3 client for 'secretsmanager' with ID 139621965464160 |
| 2025-02-05 17:18:28.298 | [DEBUG] 2025-02-05T17:18:28.298Z bc916a15-8e8c-500e-899b-0a08111b5048 azul.terra Query (296 characters total): '\n            SELECT consent_group, data_modality, data_use_permission, datarepo_row_id, dataset_id, owner, principal_investigator, registered_identifier, source_datarepo_row_ids, title\n            FROM `datarepo-dev-e53e74aa.ANVIL_1000G_2019_Dev_20230609_ANV5_202306121732.anvil_dataset`\n        ' |
| 2025-02-05 17:18:29.442 | [DEBUG] 2025-02-05T17:18:29.442Z bc916a15-8e8c-500e-899b-0a08111b5048 azul.terra Job info: {"job_id": "8a19fc25-da19-408e-b038-80eaf2e15598", "total_rows": 1, "stats": {"totalBytesProcessed": "0", "totalBytesBilled": "0", "cacheHit": true, "searchStatistics": {"indexUsageMode": "UNUSED", "indexUnusedReasons": [{"code": "QUERY_CACHE_HIT", "message": "Search indexes are not used because the query was cached."}]}}, "query": "\n            SELECT consent_group, data_modality, data_use_permission, datarepo_row_id, dataset_id, owner, principal_investigator, registered_identifier, source_datarepo_row_ids, title\n            FROM `datarepo-dev-e53e74aa.ANVIL_1000G_2019_Dev_20230609_ANV5_202306121732.anvil_dataset`\n        "} |
| 2025-02-05 17:18:29.622 | [INFO] 2025-02-05T17:18:29.622Z bc916a15-8e8c-500e-899b-0a08111b5048 azul.terra Making GET request to 'https://jade.datarepo-dev.broadinstitute.org/api/repository/v1/snapshots/790795c4-49b1-4ac8-a060-207b92ea08c5' |
| 2025-02-05 17:18:29.622 | [DEBUG] 2025-02-05T17:18:29.622Z bc916a15-8e8c-500e-899b-0a08111b5048 azul.terra … without request body |
| 2025-02-05 17:18:30.180 | [INFO] 2025-02-05T17:18:30.180Z bc916a15-8e8c-500e-899b-0a08111b5048 azul.terra Got 200 response after 0.558s from GET to https://jade.datarepo-dev.broadinstitute.org/api/repository/v1/snapshots/790795c4-49b1-4ac8-a060-207b92ea08c5 |
| 2025-02-05 17:18:30.180 | [DEBUG] 2025-02-05T17:18:30.180Z bc916a15-8e8c-500e-899b-0a08111b5048 azul.terra … with response headers HTTPHeaderDict({'Date': 'Wed, 05 Feb 2025 17:18:30 GMT', 'Server': 'Apache', 'X-Frame-Options': 'SAMEORIGIN', 'Access-Control-Allow-Headers': 'DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Authorization,Accept,Referer,X-App-Id,Origin', 'Access-Control-Allow-Methods': 'GET,POST,DELETE,PUT,PATCH,OPTIONS,HEAD', 'X-Content-Type-Options': 'nosniff', 'Strict-Transport-Security': 'max-age=31536000;includeSubDomains', 'Cache-Control': 'no-cache,no-store,must-revalidate', 'X-Request-ID': 'eObjGrzz', 'Content-Type': 'application/json', 'Content-Length': '21219', 'Vary': 'Accept-Encoding,Origin', 'Via': '1.1 google', 'Alt-Svc': 'h3=":443"; ma=2592000,h3-29=":443"; ma=2592000'}) |
| 2025-02-05 17:18:30.180 | [DEBUG] 2025-02-05T17:18:30.180Z bc916a15-8e8c-500e-899b-0a08111b5048 azul.terra … with response body b'{"id":"790795c4-49b1-4ac8-a060-207b92ea08c5","name":"ANVIL_1000G_2019_Dev_20230609_ANV5_202306121732","description":"Full vie...' |
| 2025-02-05 17:18:30.183 | [WARNING] 2025-02-05T17:18:30.181Z bc916a15-8e8c-500e-899b-0a08111b5048 azul.indexer.index_controller Worker failed to handle message {'action': 'add', 'notification': {'transaction_id': '4b3447c7-0826-4fb3-8d8c-445b9c297da7', 'bundle_fqid': {'uuid': '677dd55c-3fa3-ab07-8c98-985d94d7577e', 'version': '2022-06-01T00:00:00.000000Z', 'source': {'id': '790795c4-49b1-4ac8-a060-207b92ea08c5', 'spec': 'tdr:bigquery:gcp:datarepo-dev-e53e74aa:ANVIL_1000G_2019_Dev_20230609_ANV5_202306121732:/0'}, 'table_name': 'anvil_dataset', 'batch_prefix': None}}, 'catalog': 'anvil'}. Traceback (most recent call last):   File "/var/task/azul/indexer/index_controller.py", line 173, in contribute     contributions, replicas = self.transform(catalog, notification, delete)                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^   File "/var/task/azul/indexer/index_controller.py", line 225, in transform     bundle = service.fetch_bundle(catalog, bundle_fqid)              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^   File "/var/task/azul/indexer/index_service.py", line 201, in fetch_bundle     return plugin.fetch_bundle(bundle_fqid)            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^   File "/var/task/azul/plugins/repository/tdr.py", line 196, in fetch_bundle     bundle = self._emulate_bundle(bundle_fqid)              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^   File "/var/task/azul/plugins/repository/tdr_anvil/__init__.py", line 333, in _emulate_bundle     return self._duos_bundle(bundle_fqid)            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^   File "/var/task/azul/plugins/repository/tdr_anvil/__init__.py", line 494, in _duos_bundle     duos_id, duos_info = self.tdr.get_duos(bundle_fqid.source)                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^   File "/var/task/azul/terra.py", line 666, in get_duos     duos_id = json_str(json_dict(body['duosFirecloudGroup'])['duosId'])                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^   File "/var/task/azul/types.py", line 75, in json_dict     assert isinstance(v, dict), type(v)            ^^^^^^^^^^^^^^^^^^^ AssertionError: <class 'NoneType'> |
| 2025-02-05 17:18:30.189 | END RequestId: bc916a15-8e8c-500e-899b-0a08111b5048 |
| 2025-02-05 17:18:30.189 | REPORT RequestId: bc916a15-8e8c-500e-899b-0a08111b5048 Duration: 2076.56 ms Billed Duration: 2077 ms Memory Size: 4096 MB Max Memory Used: 149 MB Init Duration: 2588.26 ms |
---

## Formatted traceback
```
Traceback (most recent call last):
  File "/var/task/azul/indexer/index_controller.py", line 173, in contribute
    contributions, replicas = self.transform(catalog, notification, delete)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/indexer/index_controller.py", line 225, in transform
    bundle = service.fetch_bundle(catalog, bundle_fqid)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/indexer/index_service.py", line 201, in fetch_bundle
    return plugin.fetch_bundle(bundle_fqid)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/plugins/repository/tdr.py", line 196, in fetch_bundle
    bundle = self._emulate_bundle(bundle_fqid)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/plugins/repository/tdr_anvil/__init__.py", line 333, in _emulate_bundle
    return self._duos_bundle(bundle_fqid)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/plugins/repository/tdr_anvil/__init__.py", line 494, in _duos_bundle
    duos_id, duos_info = self.tdr.get_duos(bundle_fqid.source)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/terra.py", line 666, in get_duos
    duos_id = json_str(json_dict(body['duosFirecloudGroup'])['duosId'])
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/types.py", line 75, in json_dict
    assert isinstance(v, dict), type(v)
           ^^^^^^^^^^^^^^^^^^^
AssertionError: <class 'NoneType'>
```

--- a/src/azul/terra.py
+++ b/src/azul/terra.py
@@ -664,7 +664,7 @@ class TDRClient(SAMClient):
         body = self._retrieve_source(source)
         try:
             duos_id = json_str(json_dict(body['duosFirecloudGroup'])['duosId'])
-        except (KeyError, TypeError):
+        except (KeyError, AssertionError):
             log.warning('No DUOS ID available for %r', source.spec)
             return None, None
         else:

--- a/test/azul_test_case.py
+++ b/test/azul_test_case.py
@@ -5,6 +5,7 @@ from abc import (
 from collections.abc import (
     Set,
 )
+import contextlib
 from contextlib import (
     AbstractContextManager,
 )
@@ -16,6 +17,7 @@ from re import (
     escape,
 )
 from typing import (
+    Iterable,
     Optional,
 )
 from unittest import (
@@ -218,6 +220,13 @@ class AzulTestCase(TestCase):
             cleanup = instance.stop
         instance.start()
         self.addCleanup(cleanup)
+
+    @contextlib.contextmanager
+    def stacked_patches(self, patches: Iterable[patch]):
+        with contextlib.ExitStack() as context:
+            for cm in patches:
+                context.enter_context(cm)
+            yield
 
 
 class AlwaysTearDownTestCase(TestCase):

--- a/test/indexer/test_anvil.py
+++ b/test/indexer/test_anvil.py
@@ -9,6 +9,7 @@ from operator import (
     itemgetter,
 )
 from typing import (
+    Iterable,
     Type,
     cast,
 )
@@ -49,6 +50,10 @@ from azul.plugins.repository.tdr_anvil import (
 from azul.terra import (
     TDRClient,
 )
+from azul.types import (
+    JSONs,
+    MutableJSONs,
+)
 from azul_test_case import (
     TDRTestCase,
 )
@@ -68,37 +73,37 @@ def setUpModule():
 
 class DUOSTestCase(TDRTestCase, ABC):
 
-    @classmethod
-    def setUpClass(cls) -> None:
-        super().setUpClass()
-        cls._patch_duos()
+    def _mock_normal_duos(self):
+        for p in self._duos_patches(self.normal_response_bodies):
+            self.addPatch(p)
 
-    mock_duos_url = furl('https://mock_duos.lan')
+    @property
+    def normal_response_bodies(self) -> MutableJSONs:
+        duos_id = 'DUOS-000000'
+        return [
+            # TDR's /snapshots/{snapshot_id} response:
+            {
+                'name': self.source.spec.name,
+                'duosFirecloudGroup': {'duosId': duos_id}
+            },
+            # DUOS' /dataset/registration/{duos_id}:
+            {
+                'consentGroups': [{'datasetIdentifier': duos_id}],
+                'studyDescription': 'Study description from DUOS'
+            }
+        ]
 
-    duos_id = 'DUOS-000000'
-    duos_description = 'Study description from DUOS'
-
-    @classmethod
-    def _patch_duos(cls) -> None:
-        cls.addClassPatch(patch.object(type(config),
-                                       'duos_service_url',
-                                       new=PropertyMock(return_value=cls.mock_duos_url)))
-        cls.addClassPatch(patch.object(TDRClient,
-                                       '_request',
-                                       side_effect=[
-                                           Mock(spec=HTTPResponse, status=200, data=json.dumps({
-                                               'name': cls.source.spec.name,
-                                               'duosFirecloudGroup': {
-                                                   'duosId': cls.duos_id
-                                               }
-                                           })),
-                                           Mock(spec=HTTPResponse, status=200, data=json.dumps({
-                                               'consentGroups': [{
-                                                   'datasetIdentifier': cls.duos_id
-                                               }],
-                                               'studyDescription': cls.duos_description
-                                           }))
-                                       ]))
+    def _duos_patches(self, bodies: JSONs) -> Iterable[patch]:
+        responses = [
+            Mock(spec=HTTPResponse, status=200, data=json.dumps(body))
+            for body in bodies
+        ]
+        mock_url = PropertyMock(return_value=furl('https://mock_duos.lan'))
+        patches = [
+            patch.object(type(config), 'duos_service_url', new=mock_url),
+            patch.object(TDRClient, '_request', side_effect=responses)
+        ]
+        return patches
 
 
 class AnvilIndexerTestCase(AnvilCannedBundleTestCase, IndexerTestCase):
@@ -173,6 +178,7 @@ class TestAnvilIndexer(AnvilIndexerTestCase,
                         self.index_service.delete_indices(self.catalog)
 
     def test_list_and_fetch_bundles(self):
+        self._mock_normal_duos()
         source_ref = self.source
         self._make_mock_tdr_tables(source_ref)
         canned_bundle_fqids = [

--- a/test/indexer/test_anvil.py
+++ b/test/indexer/test_anvil.py
@@ -218,6 +218,35 @@ class TestAnvilIndexer(AnvilIndexerTestCase,
                 self.assertEqual(canned_bundle.links, bundle.links)
                 self.assertEqual(canned_bundle.orphans, bundle.orphans)
 
+    def test_absent_duos_id(self):
+        source_ref = self.source
+        self._make_mock_tdr_tables(source_ref)
+        cases = {
+            'Absent duosFirecloudGroup': [
+                {'name': self.source.spec.name}
+            ],
+            'Empty duosFirecloudGroup': [
+                {
+                    'name': self.source.spec.name,
+                    'duosFirecloudGroup': {}
+                }
+            ],
+            'Null duosId': [
+                {
+                    'name': self.source.spec.name,
+                    'duosFirecloudGroup': {'duosId': None}
+                }
+            ]
+        }
+        for sub_test, response_bodies in cases.items():
+            with self.subTest(sub_test):
+                with self.stacked_patches(self._duos_patches(response_bodies)):
+                    plugin = self.plugin_for_source_spec(source_ref.spec)
+                    bundle = plugin.fetch_bundle(self.duos_bundle())
+                    self.assertIsInstance(bundle, TDRAnvilBundle)
+                    self.assertEqual({}, bundle.entities)
+                    self.assertEqual(1, len(bundle.orphans))
+
 
 class TestAnvilIndexerWithIndexesSetUp(AnvilIndexerTestCase):
     """

--- a/test/indexer/test_anvil.py
+++ b/test/indexer/test_anvil.py
@@ -73,7 +73,7 @@ class DUOSTestCase(TDRTestCase, ABC):
         super().setUpClass()
         cls._patch_duos()
 
-    mock_duos_url = furl('https:://mock_duos.lan')
+    mock_duos_url = furl('https://mock_duos.lan')
 
     duos_id = 'DUOS-000000'
     duos_description = 'Study description from DUOS'

